### PR TITLE
Adding the Table portion of Ranking Status Page

### PR DESCRIPTION
**Pull request by amyleblond**
Changes:

**ranking_status.html** : adding front end for the table. 3 different loops, 1 for course output, 1 for number of applicants and one for whether or not the submission has happened
**views.py** : import of OrderedDict for sorting dictionaries on keys, for loop for counting number of applicants, for loop for determining if there are applicants for a course ' No Applicants' , for loop for whether or not the Intructor preference has occured (loops through apps table)

Testing:
course MSCI 603 - course_id is 122 - showing as application submitted on front end
screen shot 2018-02-05 at 4 09 51 pm
screen shot 2018-02-05 at 4 10 20 pm

Backend shows that applications were submitted:
screen shot 2018-02-05 at 4 10 52 pm

### CODE REVIEW

File: taworks/taform/templates/taform/ranking_status.html [hunk 1]
```diff
@@ -4,15 +4,44 @@
 <form>
     <button class="button" formaction="{% url 'home' %}">Home</button>
 </form>
-
+{% csrf_token %}
```
Comment (j2kan) [inline taworks/taform/templates/taform/ranking_status.html]: don't think we need this

File: taworks/taform/templates/taform/ranking_status.html [hunk 2]
```diff
@@ -4,15 +4,44 @@
...
 <div id="font">{{success}}</div>
 
 <h4>Below is a listing of the status of graduate student rankings for the upcoming term's courses.</h4> 
 <p>Rankings that have been submitted can still be edited by the AC. Click on the course link to view and edit Instructor rankings.</p>
 
-<h2>***TODO Ranking status displayed here***</h2> 
+<table class="tg">
+	<tr>
+	<th class="tg-8fvv">Course</th>
+	<th class="tg-8fvv">Number of Applicants</th>
```
Comment (j2kan) [inline taworks/taform/templates/taform/ranking_status.html]: to save space i think, `# of apps` is better

**Comment by j2kan** [comment]
Can you pull latest master into your branch and fix the merge conflict? I can review it fully after.

File: taworks/taform/templates/taform/ranking_status.html [hunk 3]
```diff
@@ -12,7 +12,36 @@ <h1>Ranking Status</h1>
...
+	<th class="tg-8fvv">Instructor</th>
+	<th class="tg-8fvv">Instructor Email</th>
+	<th class="tg-8fvv">Ranking Status</th>
+	</tr>
+{% for c in courses %}
+<tr>
+	<td class="tg-yw4l"><a target="_blank" href="https://team4.uwaterloo.ca/taform/instructor/{{c.url_hash}}/">{{c.course_subject}}&nbsp;{{c.course_id}}</a></td>
+	<td>
+		{% for key,value in num_applicants.items %}
+			{% if forloop.parentloop.counter == forloop.counter %}
```
Comment (j2kan) [inline taworks/taform/templates/taform/ranking_status.html]: Are you able to preprocess the 3 things: course, # of applicants, ranking_status
as 1 object in views.py?

So we don't need to run 2 nested for loops to render the page here and hopefully just do.

c.course_id, c.num_apps, c.rank_status in 1 shot.

**Comment by j2kan** [comment]
this is a spot I'd recommend raw sql since you're doing more complicated aggregation with conditioning preference != 0 and I think it would be a lot easier to process from there.

I believe your raw sql can return you evreything you need.
course, # apps, ranking status based on a case statement.

**Comment by smwatts** [comment]
I disagree with the use of raw SQL, I think we can find a nice way to do this with pandas dfs.

For example in the the export function I do all my data manipulation in views then export the results. 
This includes joins across 3 tables: students, courses and applications and cleaning up the data so its in the format that you want. For example you can make a new df with the counts by doing something like df = df.groupby('some_column')

I also join across the tables using the "merge" function. -> Anything you want to do in SQL can be done using pandas and its clean if you prep a data frame in views then pass it into your html page and just output line by line.

Here is my example:
https://github.com/j2kan/ta-works/pull/39
screen shot 2018-02-05 at 9 24 09 pm

I am happy to look into this with in more detail if you like @amyleblond. Just quick suggestions based on @j2kan's comment

To render the dataframe as a table we could try something like this: https://stackoverflow.com/questions/39003732/display-django-pandas-dataframe-in-a-django-template

Not exactly sure how it'll work with our CSS & with the hyperlinks to courses but would be cool!

**Comment by j2kan** [comment]
Do we also want to display the section?

**Comment by j2kan** [comment]
@smwatts when I said raw sql. I sort of meant using django built in aggregate functions and joins.
https://docs.djangoproject.com/en/2.0/topics/db/aggregation/

**Comment by smwatts** [comment]
Ohh thats very clean @j2kan, sorry thought you meant RAW SQL. @amyleblond definitely look at this first

**Comment by amyleblond** [comment]
updates :) Thanks @j2kan 

screen shot 2018-02-06 at 6 40 43 pm


screen shot 2018-02-06 at 6 42 02 pm

**Review by smwatts — COMMENTED** [review]
@amyleblond this code is so beautiful

File: taworks/taform/views.py [hunk 4]
```diff
@@ -39,7 +40,27 @@ def ranking_status(request):
...
+        avgRating = Avg('instructor_preference')))
+
+    for r in ranking_status:
+        if(r['count']==0):
+            r['status']='No Applicants'
+        elif(r['avgRating'] is None):
+            r['status']='Not Submitted'
+        else:
+            r['status']='Submitted'
+    print ranking_status
```
Comment (j2kan) [inline taworks/taform/views.py]: remove this line

**Review by j2kan — COMMENTED** [review]
just remove the print otherwise LGTM

**Review by j2kan — APPROVED** [review]
lgtm

**Pull request merged**